Fix collision between CONFIG_POLL and CONFIG_COHERENCE

andyross opened the pull request:
This really isn't a complicated PR.  All it does it put the data that used to be in the struct _poller tracking struct on the caller stack in k_poll() into the thread struct, in the process shrinking it down to just two bytes.  But the surgery was a little complicated due to some feature creep, so it's split out into multiple steps for review.

Commit:
kernel/poll: Make the poll callback a single-byte enum instead

Somewhat weirdly, k_poll() can do one of two things when the set of
events is signaled: it can wake up a sleeping thread, or it can submit
an unrelated work item to a work queue.  The difference in behaviors
is currently captured by a callback, but as there are only two it's
cleaner to put this into a "mode" enumerant.  That also shrinks the
size of the data such that the poller struct can be moved somewhere
other than the calling stack.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/poll: Use a separate thread backpointer for triggered work

The triggered work scheme uses a trick where it overloads the thread
pointer field of the struct poller (which normally stores the ID of
the thread that is blocked in k_poll()) to be able to find the work
queue to which it will submit.

Give it its own pointer field to break this false dependency.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/poll: Store the work queue in the triggered work poll items

The poll code was playing this weird trick where the thread pointer in
the "struct _poller" object for a triggered work item.  It would not
be a thread to wake up, but instead a pointer to the (non-polling)
thread operated by the work queue being triggered.  The code would
never touch this thread, just use it as a way to get a pointer to the
enclosing work queue struct.

Just store the work queue pointer in the first place.  It's much
simpler, and makes future modifications to remove that thread pointer
possible.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/poll: Move "_poller" into thread struct, fix COHERENCE collision

Fix the issue where the kernel poll code would place the tracking
struct on the caller stack and share it with other threads, thus
creating a cache coherence issue on systems where KERNEL_COHERENCE is
enabled.

This works by eliminating the thread backpointer in struct _poller and
simply placing the (now just two-byte!) struct directly into the
thread struct.

Note that this doesn't attempt to fix the API paradigm that the
natural way to structure a call to k_poll() is to use an array of
k_poll_events on the CALLER's stack.  So it's likely that most
"typical" k_poll code is still going to have problems with
KERNEL_COHERENCE.  But at least now the kernel internals aren't
fundamentally broken.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/queue: Remove fifo/queue "poll" variant tests

These test variants were there to test an older backend to the kernel
queue utility that used k_poll() as the blocking mechanism.  That code
got removed a while back, so these tests were just dupicates of the
main cases now.  Remove.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests: Remove filtering on KERNEL_COHERENCE

These tests were suppressed when KERNEL_COHERENCE=y because of a
feature collision with CONFIG_POLL that has since been fixed.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

tyrantlizard77 reviewed (approved):
small naming nit otherwise LGTM

Commit:
kernel: Rename struct _poller

This legacy struct still had a non-standard name.  Clean it up to
conform to currrent naming guidelines.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented on include/kernel.h:
```diff
@@ -272,6 +272,14 @@ struct _thread_userspace_local_data {
+/* private, used by k_poll and k_work_poll */
+struct k_work_poll;
+typedef int (*_poller_cb_t)(struct k_poll_event *event, uint32_t state);
+struct _poller {
```
Fair enough.  It was an older struct that just got moved, but easy fix.
